pkg/server/distributor_test.go

```diff
@@ -0,0 +1,475 @@
+			require.GreaterOrEqual(t, count, 1, "instance did not get any traffic: "+instance)
+		}
+	})
+
+	stopErrs := make(map[string]error)
+	stopServers := func(done chan error) {
+		var wg sync.WaitGroup
+		for _, testServer := range testServers {
+			wg.Add(1)
+			go func(s testModuleServer) {
+				defer wg.Done()
+				ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+				defer cancel()
+				if err := testServer.server.Shutdown(ctx, "tests are done"); err != nil {
+					stopErrs[testServer.id] = err
+				}
+			}(testServer)
+		}
+		wg.Wait()
+
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		defer cancel()
+		if err := distributorServer.server.Shutdown(ctx, "tests are done"); err != nil {
+			stopErrs[distributorServer.id] = err
+		}
+		done <- nil
+	}
+
+	done := make(chan error, 1)
+	go stopServers(done)
+	select {
+	case <-done:
+	case <-time.After(30 * time.Second):
+		t.Fatal("timeout waiting for servers to shutdown")
+	}
+
+	for server, runErr := range runErrs {
+		t.Fatalf("unexpected run error from module server %s: %v", server, runErr)
+	}
+
+	for server, stopErr := range stopErrs {
+		t.Fatalf("unexpected stop error from module server %s: %v", server, stopErr)
+	}
+}
+
+func startAndWaitHealthy(t *testing.T, testServer testModuleServer, runErrs map[string]error) {
+	go func() {
+		// this next line is to avoid double registration, as both InitializeDocumentBuilders as well as ProvideUnifiedStorageGrpcService
+		// are hard-coded to use prometheus.DefaultRegisterer
+		// the alternative would be to get the registry from wire, in which case the tests would receive a new
+		// registry automatically, but that _may_ change metric names
+		// We can remove this once that's fixed
+		prometheus.DefaultRegisterer = prometheus.NewRegistry()
+		if err := testServer.server.Run(); err != nil && !errors.Is(err, context.Canceled) {
+			runErrs[testServer.id] = err
+		}
+	}()
+
+	deadline := time.Now().Add(20 * time.Second)
+	for {
+		require.NoError(t, runErrs[testServer.id], "failed to start "+testServer.id)
+
+		conn, err := net.DialTimeout("tcp", testServer.grpcAddress, 1*time.Second)
+		if err == nil {
+			_ = conn.Close()
+			break
+		}
+
+		if time.Now().After(deadline) {
+			t.Fatal("server failed to become ready: ", testServer.id)
+		}
+
+		time.Sleep(1 * time.Second)
+	}
+
+	res, err := testServer.healthClient.Check(context.Background(), &grpc_health_v1.HealthCheckRequest{})
+	require.NoError(t, err)
+	require.Equal(t, res.Status, grpc_health_v1.HealthCheckResponse_SERVING)
+}
+
+type testModuleServer struct {
+	server         *ModuleServer
+	healthClient   grpc_health_v1.HealthClient
+	resourceClient resource.ResourceClient
+	id             string
+	grpcAddress    string
+}
+
+func initDistributorServerForTest(t *testing.T) testModuleServer {
+	cfg := setting.NewCfg()
+	cfg.HTTPPort = "13000"
+	cfg.GRPCServer.Network = "tcp"
+	cfg.GRPCServer.Address = "127.0.0.1:20000"
```

Comment:
Same comment about port assignment -- we should let the system assign the random unique ports, and then find out what it used.